### DISCUSSION TITLE
Lazy Loading properties once dependent property is set async

### DISCUSSION THREAD

**Discussion by clodewyks** [MarimerLLC/csla#3073] [Questions] [answered]
We are busy migrating our application from CSLA 4.11.2 to CSLA 6.x and WPF to Blazor.

An issue that we are currently running into is the following:

We have `Lookups` (Key value pairs) on our object that can only be loaded once a dependent property is set, for example:

    [Serializable]
    public class Plan : BusinessBase<Plan>
    {

        public static readonly PropertyInfo<string> PlanTypeProperty = RegisterProperty<string>(c => c.PlanType);
        public string PlanType
        {
            get { return GetProperty(PlanTypeProperty); }
            set { SetProperty(PlanTypeProperty, value); }
        }

         //Can only be determined once we have the PlanType
        public static readonly PropertyInfo<Lookup> PricingOptionsProperty = RegisterProperty<Lookup>(c => c.PricingOptions);
        public Lookup PricingOptions
        {
            get { return GetProperty(PricingOptionsProperty); }
            set { SetProperty(PricingOptionsProperty, value); }
        }
    
    }

In the past, and without the requirement of async that Blazor Webassembly enforces, we would handle this using the OnPropertyChanged:

        protected override void OnPropertyChanged(string propertyName)
        {
            if (propertyName == nameof(PlanType))
                SetProperty(PricingOptions, Lookups.GetPricingOptions(PlanType));
      }

With the requirement of async, and OnPropertyChanged being sync, this is no longer a viable option. We have, against our own better sense and Rocky's many warnings, tried using an async rule and the outcome was unfavorable as expected.

What is the correct way to load properties based on the value of another property being set?

**Comment by Chicagoan2016** [comment]
I don't have an answer : ) but I am a little jealous! you folks get to migrate to the latest Csla version.
What are you using for Dependency injection?
Thanks

**Reply by clodewyks** [reply]
We are very excited to get started on this new project, we have used CSLA 5.x in some other smaller projects but now we are looking to overhaul our entire administration system - exciting and terrifying!
We are using .NET for dependency injection through IServiceCollection.

**Comment by danielmartind** [comment, marked as answer]
I've used an async rule to manage similar lookup properties.

**Reply by clodewyks** [reply]
@danielmartind do you have some example code?

**Reply by danielmartind** [reply]
This adapted from a lookup I have. It is CSLA 5 though so may need an update for v6

``` 
public class LookupPricing : PropertyRuleAsync
{
          private IPropertyInfo PricingOptionsProperty { get; set; }

          public LookupPricing(IPropertyInfo primaryProperty, IPropertyInfo pricingOptionsProperty) : base(primaryProperty)
          {
              PricingOptionsProperty = pricingOptionsProperty;
              (InputProperties ?? (InputProperties = new List<IPropertyInfo>())).Add(primaryProperty);
              AffectedProperties.Add(pricingOptionsProperty);
              CanRunAsAffectedProperty = false;
              CanRunInCheckRules = false;
              CanRunOnServer = false;
          }

          protected override async Task ExecuteAsync(IRuleContext context)
          {
              var planType = context.InputPropertyValues[PrimaryProperty].ToString();
              try
              {
                  var option = await Lookups.GetPricingOptions(planType);
                  context.AddOutValue(PricingOptionsProperty, option);

              }
              catch (Exception ex)
              {
                  context.AddErrorResult(ex.Message);
              }                
          }
}
```

In add business rules
`BusinessRules.AddRule(new LookupPricing (PlanTypeProperty , PricingOptionsProperty));`


I think I used [this](https://github.com/MarimerLLC/csla/blob/b4d88567683729471103966e030fff4fbcbfb52e/Samples/RuleTutorial/AsyncLookupRule/Rules/AsyncLookupCustomer.cs) as a basis.

**Reply by clodewyks** [reply]
My first born child will be named "@danielmartind", future generations will sing songs about how the original @danielmartind saved my bacon! Thank you! It appears the essential part I was missing after working through the sample rules was `context.AddOutValue`. The solution works great.

**Reply by Chicagoan2016** [reply]
@clodewyks Did you have to change any code since @danielmartind reply was for Csla 5.xx?

A fun fact about @danielmartind ! he is from Lancashire, UK. Lancashire county has a great cricket team.
I have been told nothing beats watching a Cricket game at Old Trafford, Manchester! gotta try it one day!!!

**Comment by clodewyks** [comment]
Often the answer is (possibly) in the question 😆 

`LazyGetPropertyAsync` appears to be what we are looking for here, when `PricingOptions` is fetched, we call an async method to check that the `Plan` is set, in which case, we return the values for this `Plan`, if not set we return an empty list. Unsure whether this is the correct approach and would still like to hear some ideas.

Idea came from the following discussion: https://github.com/MarimerLLC/csla/discussions/2094

Will keep you posted.
